fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/AutoContinuingCursor.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * AutoContinuingCursor.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2019 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.cursors;
+
+import com.apple.foundationdb.API;
+import com.apple.foundationdb.record.RecordCoreException;
+import com.apple.foundationdb.record.RecordCursor;
+import com.apple.foundationdb.record.RecordCursorResult;
+import com.apple.foundationdb.record.RecordCursorVisitor;
+import com.apple.foundationdb.record.provider.foundationdb.FDBDatabaseRunner;
+import com.apple.foundationdb.record.provider.foundationdb.FDBRecordContext;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.NoSuchElementException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.function.BiFunction;
+
+/**
+ * A wrapper for a cursor to iterate across multiple transactions.
```

Comment:
I think you need more description here (you had a comment in your PR about adding javadocs, so maybe this is intentional).  This should include a detailed description of exactly how it works and any restrictions it has.  For example: 

* It is provided a generator that produces a cursor
* The cursor is iterated over until it is either exhausted or until scan limit properties are reached
   * if scan limit properties are reached, it re-starts the cursor in the context of a new transaction
* Once a cursor is exhausted, the generator is asked for the next cursors and the process is repeated
* It is responsible for all `FDBRecordContext` management, so all reads it does are in the scope of transactions that it controls.
* The cursor has no visibility into the `ScanProperties` of cursors returned by the generator and, therefore, will not be involved in enforcing any limits that may be, individually, applied to the `ScanProperties` that may be applied to the cursor returned by the generator.  For example, if the generator returns a cursor that specified, say, a record scan limit of 10 record, the `AutoContinuingCursor` will scan all data until it is exhausted, 10 records at a time.

Reply:
Yes, I had a weird mental hiccup with what the cursor was supposed to do :-) 

However, it is *possible* that the user wanted to provide a cursor generator that chained together different cursors.  I.e., use cursor A until exhausted, then use cursor B until exhausted. So just clarifying that once a returned cursor declares its source is exhausted, then the `AutoContinuingCursor` is exhausted.

Reply:
Yes. See `If the <i>underlying cursor</i> is exhausted, the {@link AutoContinuingCursor} is also exhausted.`